### ISSUE TITLE
204/205 responses should discard content instead of throwing

### ISSUE THREAD

**Issue by tadfisher** [square/okhttp#6836]
While HTTP gives some leeway in how user-agents can handle response bodies for 204 and 205 responses, no major Web browser (Chrome, Firefox, Safari) treats such responses as errors. OkHttp does, however: https://github.com/square/okhttp/blob/master/okhttp/src/main/kotlin/okhttp3/internal/http/CallServerInterceptor.kt#L138-L141

This wouldn't ordinarily be a problem as we can rewrite responses, except this error is being thrown in `CallServerInterceptor` and is the last interceptor in the chain. So as a client, there is no reliable way to know that the response succeeded without matching against the message text of `ProtocolException`.

Ideally, OkHttp would behave as the browsers do and simply discard these response bodies, even if they are erroneous per the spec.

**Comment by yschimke** [comment]
Are there known servers that produce this? Or an error from internal software?

**Comment by yschimke** [comment]
This also seems awkward, it would silently consume the body to keep the connection working.  Closing out as won't fix, unless we can see this as a wider problem.

@tadfisher Are there well known servers that produce this, or is it typically a misconfigured personal server?

**Comment by rfc2822** [comment]
For the records: We (DAVx5 CalDAV/CardDAV client for Android) get this problem (server returns 204 plus some content, usually a positive status message) occasionally. We always tell people to ask their server admins to fix it. If they can reach somebody, it will be fixed in most cases. However this is not user-friendly.

I would hope that the content of 204 responses would either be ignored by default or at least that we could catch the error in our code (specific ProtocolException or something like that).

Is there any disadvantage of discarding the body? I think it would be correct in the sense of the Robustness principle.

**Comment by Gupta-Aryaman** [comment]
@yschimke @swankjesse 
After several years, I’m encountering this issue again. The server is returning a body in the first 204 response, which eventually leads to a ProtocolException when parsing the subsequent request.

I understand that reading (and skipping) the body of a 204 response would violate HTTP semantics. However, it seems that the more appropriate behaviour would be to fail immediately on the first 204 response that incorrectly contains a body, rather than allowing it to affect the parsing of the next request/response on the connection.

It might be worth re-evaluating this behaviour. I’d be happy to open a PR to address this if it makes sense.

I am using okhttp library's 4.12.0 version.

**Comment by yschimke** [comment]
Can you implement a network interceptor to detect this and mark the connection closed?

**Comment by Gupta-Aryaman** [comment]
Sure I will try. Could you please assign this issue to me?

**Comment by yschimke** [comment]
The interceptor should be in your project, not okhttp

**Comment by Gupta-Aryaman** [comment]
That is one solution to approach this problem, sure. But don't you think, as a http library the error should come on the same http call instead of messing up the next call? If not, what's your opinion in the contrary?

**Comment by swankjesse** [comment]
@Gupta-Aryaman how would we detect that? I believe that the RFCs tell us the first call is complete, and therefore the bytes that follow are from the next call.

**Comment by Gupta-Aryaman** [comment]
I was thinking that if we receive a status code of 204 or 304 with a content-length of > 0 then mark that request as malformed or something. 

Similar to this [PR](https://github.com/square/retrofit/pull/815/changes/0152ffa1f923a23b2e2cd658a7141d346db132c8) in retrofit

But I can see that you had a different opinion there as well 😄

**Comment by Gupta-Aryaman** [comment]
If you feel that throwing an error is on such a response is not what client should experience, then the library could also see the content-length on such responses and ignore the specific number of bytes before starting to parse the next response. But in any case, I believe that next request should not suffer because of a unexpected response from the previous request.

**Comment by yschimke** [comment]
The server is broken. You should be filing this with them, rather all potential clients
